// @shoutem/ui/components/RichMedia/lib/LazyRender.js
import React from 'react';

import {
  View,
  InteractionManager,
  Text,
} from 'react-native';

function DefaultPlaceholder() {
  return (
    <View>
      <Text>Loading...</Text>
    </View>
  );
}

export default function (renderPlaceholder) {
  return function lazyRender(WrappedComponent) {
    /**
     * When used on a React Component adds lazy rendering
     *
     */
    return class LazyRender extends WrappedComponent {
      state = {
        isReady: false,
      }

      componentDidMount() {
        InteractionManager.runAfterInteractions(() => {
          this.setState({ isReady: true });
        });

        if (super.componentDidMount) {
          super.componentDidMount();
        }
      }

      renderPlaceholder() {
        if (renderPlaceholder) {
          return renderPlaceholder();
        }

        return <DefaultPlaceholder />;
      }

      render() {
        if (!this.state.isReady) {
          return this.renderPlaceholder();
        }

        return <WrappedComponent {...this.props} isReady={this.state.isReady} />;
      }
    };
  };
}
